### PR TITLE
Refactor Exposures

### LINKED ISSUE #12: Disease as an object
Agents should carry diseases as an object rather than simply booleans. Could greatly enhance the functionality around types of disease, severity, etc.

### LINKED ISSUE #94: Make random trial generic
Currently, random trials are linked to PrEP and only work in the case where features.prep is on. They can be used for assigning PrEP or peer-change agents, but should be abstracted.

### PR DESCRIPTION
**Pull request by mcmcgrath13**
* Move exposures into `exposures/` directory with all logic for the exposure consolidated there
* Similar to `features` - might need to think about how to consolidate further the common parts
* Questions:
  * Is it okay to have agents always interact (who is agent/partner arbitrary at that point), then go through each exposure to determine if that interaction changes anything?
 
To Do:
- [x] knowledge
- [x] unit tests
- [x] docs

closes #12, closes #94

### CODE REVIEW

**Review by s-bessey — COMMENTED** [review]
I think you said this is still in the works, so I just looked generally. I like the direction!

File: titan/exposures/knowledge.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+        args:
+            pop: the population this agent is a part of
+            time: the current time step
+        """
+        knowledge_params = self.agent.location.params.knowledge
+        if pop.pop_random.random() < knowledge_params.init:
+            self.active = True
+
+            # TO_REVIEW - non aware agents always start at opinoin of 0?
```
Comment (s-bessey) [inline titan/exposures/knowledge.py:43]: Is this still to review?

File: titan/features/random_trial.py [hunk 2]
```diff
@@ -14,21 +14,21 @@ def __init__(self, agent):
 
         self.active = False
         self.treated = False
+        self.suitable = False
 
     @classmethod
-    def update_pop(cls, model: "model.HIVModel"):
+    def update_pop(cls, model: "model.TITAN"):  # TODO rename TITAN to TITAN
```
Comment (s-bessey) [inline titan/features/random_trial.py:20]: Remove TODO

File: titan/population_io.py [hunk 3]
```diff
@@ -42,18 +42,14 @@
 )
 
 
-def write(
-    pop: Population, dir: str, intervention_attrs: bool = False, compress: bool = True
-) -> str:
+# TO DO get rid of option to only write intervention attrs
```
Comment (s-bessey) [inline titan/population_io.py:45]: Is this to-do done? It looks like it

**Review by s-bessey — COMMENTED** [review]
I left single comments instead of a review, but just a couple of clean-up things.

File: titan/exposures/knowledge.py (hunk 1 above)
Comment (mcmcgrath13) [reply titan/exposures/knowledge.py:43]: yeah

Comment (s-bessey) [reply titan/exposures/knowledge.py:43]: I don't think it makes a difference about whether they get an opinion at initialization or when they become aware model-wise but it makes more sense to initiate with 0 from a non-model perspective?

**Comment by mcmcgrath13** [comment]
@s-bessey in doing the merge and cleaning up the tests I found a few things that need reviewing (including the one causing the unit tests to fail) - could you take a look?

File: docs/api/interactions/index.md [hunk 4]
```diff
@@ -12,6 +12,6 @@ To add a new interaction:
 * Add tests in `tests/interactions/`
 * Add it to the docs in `docs/api/interactions/` and to the nav in `mkdocs.yml`
 
-The `HIVModel` class uses sub-classes of `BaseInteraction` to initialize the object/call methods as appropriate.
+The `TITAN` class uses sub-classes of `BaseInteraction` to initialize the object/call methods as appropriate.
```
Comment (s-bessey) [inline docs/api/interactions/index.md:15]: This change is so exciting!

File: titan/features/vaccine.py [hunk 5]
```diff
@@ -44,19 +44,19 @@ def init_agent(self, pop: "population.Population", time: int):
...
 
         If PrEP feature is enable and the agent is not active PrEP and not HIV, either update or stochastically vaccinate the agent.
 
         args:
-            model: the instance of HIVModel currently being run
+            model: the instance of TITAN currently being run
         """
         if (
-            model.params.features.prep
+            model.params.features.prep  # TO_REVIEW should this be required?
```
Comment (s-bessey) [inline titan/features/vaccine.py:57]: No, we should have `model.params.features.vaccine`.

File: tests/features/random_trial_test.py [hunk 6]
```diff
@@ -37,8 +107,10 @@ def test_initialize_random_trial_pca_bridge(make_model, params):
         if bridges:
             bridge_num += 1
             for agent in comp.nodes():
-                if agent.pca.active:
-                    assert agent in [ag for ags in bridges for ag in ags]
+                if agent.knowledge.active:
+                    assert agent in [
+                        ag for ags in bridges for ag in ags
+                    ]  # TO_REVIEW what?
```
Comment (s-bessey) [inline tests/features/random_trial_test.py:113]: I think what this is testing is that the chosen agents are actually in the bridging agent list it should be choosing from?

**Review by s-bessey — COMMENTED** [review]
To_reviews reviewed. Let me know if I missed anything (I only looked on github, but I think I uncollapsed all the files)

File: requirements.txt [hunk 7]
```diff
@@ -1,5 +1,4 @@
 wheel==0.34.2
-matplotlib==3.2.0
```
Comment (s-bessey) [inline requirements.txt:2]: :tada:

File: settings/chicago/chicago_baseCase.py [hunk 8]
```diff
@@ -224,9 +224,7 @@
     params.prep.discontinue = 0.00  # 0.15
     params.prep.peak_load = 1.0
     params.prep.half_life = 1.0
-elif (
```
Comment (s-bessey) [inline settings/chicago/chicago_baseCase.py:227]: This file should probably be deleted?

File: tests/exposures/hiv_test.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+from conftest import FakeRandom
+
+from titan.exposures import HIV
+from titan.agent import Relationship
+
+
+@pytest.mark.unit
+def test_hiv_expose(make_model, make_agent):
+    model = make_model()
+    model.run_random = FakeRandom(-0.1)  # forces conversion even if prob is 0
```
Comment (s-bessey) [inline tests/exposures/hiv_test.py:12]: It'd be nice to standardize the comment for `FakeRandom(-0.1)`. Below it's `# always less than param`

File: tests/features/external_exposure_test.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+    model = make_model(params)
+    external_exposure_agent = make_agent()
+    external_exposure_agent.external_exposure.active = True
+    model.pop.add_agent(external_exposure_agent)
+
+    model.run_random = FakeRandom(-0.1)
+
+    external_exposure_agent.external_exposure.update_agent(model)
+    assert external_exposure_agent.hiv.active
```
Comment (s-bessey) [inline tests/features/external_exposure_test.py:18]: Does the agent always initialize with `hiv.active = False`? I'm not sure, but if it's not sure, should set it that way.

Looks like there might be a missing newline at the end of this file.

File: tests/integration_test.py [hunk 11]
```diff
@@ -144,7 +144,7 @@ def test_target_partners(make_model_integration, tmpdir):
     model_a.params.model.seed.run = model_a.run_seed
     model_a.params.model.seed.ppl = model_a.pop.pop_seed
 
-    model_b = HIVModel(model_a.params)
+    model_b = TITAN(model_a.params)
```
Comment (s-bessey) [inline tests/integration_test.py:147]: 🎉

File: tests/interactions/injection_test.py [hunk 12]
```diff
@@ -2,42 +2,73 @@
...
 @pytest.mark.unit
 def test_injection_transmission(make_model, make_agent):
     model = make_model()
+    model.np_random = FakeRandom(1.0)
+    model.run_random = FakeRandom(-0.1)
     model.time = model.params.hiv.start_time + 2
-    a = make_agent(race="white", DU="Inj", SO="HM")
-    p = make_agent(race="white", DU="Inj", SO="HF")
+    a = make_agent(race="black", DU="Inj", SO="HM")
+    p = make_agent(race="black", DU="Inj", SO="HF")
```
Comment (s-bessey) [inline tests/interactions/injection_test.py:17]: Why did race change here?

File: titan/features/base_feature.py [hunk 13]
```diff
@@ -57,12 +57,12 @@ def init_agent(self, pop: "population.Population", time: int):
         """
         pass
 
-    def update_agent(self, model: "model.HIVModel"):
+    def update_agent(self, model: "model.TITAN"):
```
Comment (s-bessey) [inline titan/features/base_feature.py:60]: 🦖

File: titan/features/haart.py [hunk 14]
```diff
@@ -57,8 +57,8 @@ def init_agent(self, pop: "population.Population", time: int):
             .drug_type[self.agent.drug_type]
         )
         if (
-            self.agent.hiv
-            and self.agent.hiv_dx
+            self.agent.hiv.active  # type: ignore[attr-defined]
```
Comment (s-bessey) [inline titan/features/haart.py:60]: Is there any way to make mypy hate this less other than the ignore?

File: titan/params/agent_zero.yml [hunk 15]
```diff
@@ -13,9 +13,16 @@ agent_zero:
...
-    default: false
\ No newline at end of file
+    default: false
+  exposure:
+    type: enum
+    default: hiv
+    description: For which exposure the agent zero should be converted
+    values:
+      - hiv
+      - knowledge
```
Comment (s-bessey) [inline titan/params/agent_zero.yml:28]: is there a newline at the end of this file?

File: titan/population_io.py [hunk 16]
```diff
@@ -210,15 +216,15 @@ def create_agent(
         if attr not in init_attrs:
             setattr(agent, attr, eval(val))
 
-    for feature in agent_feats:
-        feat_row = agent_feats[feature][agent.id]
-        agent_feat = getattr(agent, feature)
-        for attr, val in feat_row.items():
+    for extra in agent_extras:
```
Comment (s-bessey) [inline titan/population_io.py:219]: What is an agent_extra?

File: titan/utils.py [hunk 17]
```diff
@@ -198,3 +200,32 @@ def override_param(params, param_path, value, delimiter="|"):
...
+
+
+def connected_components(graph):
+    """
+    Get connected components in graph
+
+    returns:
+        list of connected components
+    """
+    return list(graph.subgraph(c).copy() for c in nx.connected_components(graph))
```
Comment (s-bessey) [inline titan/utils.py:231]: newline

**Review by s-bessey — CHANGES_REQUESTED** [review]
Almost there! Beautiful refactor.

File: tests/features/external_exposure_test.py (hunk 10 above)
Comment (mcmcgrath13) [reply tests/features/external_exposure_test.py:18]: yeah, active always starts false on agent creation

Comment (s-bessey) [reply tests/features/external_exposure_test.py:18]: 👍 I thought so but wasn't sure if the conftest agent creation altered it or something. Which wouldn't make sense, in retrospect

File: tests/interactions/injection_test.py (hunk 12 above)
Comment (mcmcgrath13) [reply tests/interactions/injection_test.py:17]: no clue - params look the same to me

File: titan/features/haart.py (hunk 14 above)
Comment (mcmcgrath13) [reply titan/features/haart.py:60]: not that I've figured out.  At least not without hardcoding all the features/things on the agent side, which I also don't want to do.

File: titan/population_io.py (hunk 16 above)
Comment (mcmcgrath13) [reply titan/population_io.py:219]: features/exposures we staple on to agent

**Review by s-bessey — APPROVED** [review]
🚀

**Pull request merged**